### ISSUE TITLE
Problem compiling the first example from the docs.

### ISSUE THREAD

**Issue by oderwat** [lfairy/maud#1]
I tried to compile the example from the docs and get:

```
Compiling maud v0.0.1 (file:///...../maud)
main.rs:9:18: 11:7 error: borrowed value does not live long enough
main.rs:9     let markup = html! {
main.rs:10         p { "Hi, " $name "!" }
main.rs:11     };
main.rs:1:1: 14:1 note: in expansion of html!
main.rs:9:18: 11:7 note: expansion site
main.rs:7:11: 14:2 note: reference must be valid for the block at 7:10...
```
(Additionally I can't use `rustc --pretty expanded` as it seems that is gone in current nightly 2015-01-24?)

**Comment by lfairy** [comment]
Thanks for reporting that -- I can confirm it fails for me too.

What the `html!` macro does internally is create a closure, then immediately borrow that closure as a trait object (`&Fn`). I assumed the closure would live long enough for the borrow, but it turns out it doesn't.

You can work around it by moving the `.render()` call into the `let`:

```rust
let markup = html! {
    p { "Hi, " $name "!" }
}.render();
assert_eq!(markup, ...);
```

Since the `Markup` is consumed before the end of the statement, it compiles fine.

I've fixed the docs to use this workaround. Likely the next version will take the internal closure by value instead of by reference, since it's pretty confusing otherwise.

**Comment by oderwat** [comment]
Got it. Also did not expect that the closure just vanishes even if "still being used" later on. I think that will make Rust hard to understand. It is just not intuitive (until it gets it... eventually... hopefully). At least it makes more sense to me than haskell :( which I failed miserably at adapting to.

**Comment by lfairy** [comment]
Okay, I've released v0.1.1 which should fix the issue. Now the `Markup` object takes ownership of the closure, so it won't vanish any more. Again, thanks for the report!

**Comment by oderwat** [comment]
Works! Thank you too! I am eager to see how this project continues. I always liked the idea of hamlet too and think this may be something we will include in some stuff we are going to develop soon.
